### ISSUE TITLE
test data transform

### ISSUE THREAD

**Issue by meehirmhatrepy** [drprojects/superpoint_transformer#160]
hi,
I have tried to run the model my my custom dataset.
I have kept val and test dataset same.
I get good accuracy on val data but poor accuracy on test data inspite it being same.
How do I apply same data transforms to my test data so that this issue doesnt exist?
i have tried to make on_device_test_transform and test_transform same as on_device_test_transform and val_transform which doesn't work for me.
If you have any suggestions for this please do let me know.
Thankyou.

**Comment by drprojects** [comment]
The provided datasets by default apply the same transforms on val and test by setting:

```yaml
test_transform: ${datamodule.val_transform}
on_device_test_transform: ${datamodule.on_device_val_transform}
```

Combined  with setting `val_on_test: True`, you should normally get the exact same results on val and test.
 
If that is not the case, please provide a minimum reproducible example if you want us to have a look.

**Comment by meehirmhatrepy** [comment]
i dont want to do val_on_test as it still do test transformations on val which makes my val accuracy bad and by setting 
test_transform: ${datamodule.val_transform}
on_device_test_transform: ${datamodule.on_device_val_transform}

i still get different results for same data:
for test:
NAGRemoveKeys(level=all, keys=('is_val',))
NAGRemoveKeys(level=all, keys=('is_val',))
NodeSize(low=0)
NAGCast(fp_dtype=torch.float32, int_dtype=torch.int64, rgb_to_float=True)
OnTheFlyHorizontalEdgeFeatures(keys=(), use_mean_normal=False)
OnTheFlyVerticalEdgeFeatures(keys=(), use_mean_normal=False)
NAGAddKeysTo(level=0, keys=['planarity', 'scattering', 'length', 'volume', 'curvature'], to=x, delete_after=True)
NAGAddKeysTo(level=1+, keys=[], to=x, delete_after=True)
NAGAddKeysTo(level=all, keys=['rgb'], to=x, delete_after=True)
NAGAddSelfLoops()
OnTheFlyInstanceGraph(level=-1, num_classes=3, adjacency_mode=radius, k_max=30, radius=0.1, use_batch=True, centroid_mode=iou, centroid_level=1, smooth_affinity=True)
NAG(num_levels=4, num_points=[162500, 37224, 37224, 34126], device=cpu)
NAG(num_levels=4, num_points=[162500, 37224, 37224, 34126], device=cuda:0)
Data(pos=[162500, 3], curvature=[162500, 1], length=[162500, 1], planarity=[162500, 1], pos_offset=[3], rgb=[162500, 3], scattering=[162500, 1], super_index=[162500], volume=[162500, 1], sub=Cluster(num_clusters=162500, num_points=174428, device=cuda:0), x=[162500, 8], semantic_pred=[162500])

for val:
NAGRemoveKeys(level=all, keys=('is_val',))
NAGRemoveKeys(level=all, keys=('is_val',))
NodeSize(low=0)
NAGCast(fp_dtype=torch.float32, int_dtype=torch.int64, rgb_to_float=True)
OnTheFlyHorizontalEdgeFeatures(keys=(), use_mean_normal=False)
OnTheFlyVerticalEdgeFeatures(keys=(), use_mean_normal=False)
NAGAddKeysTo(level=0, keys=['planarity', 'scattering', 'length', 'volume', 'curvature'], to=x, delete_after=True)
NAGAddKeysTo(level=1+, keys=[], to=x, delete_after=True)
NAGAddKeysTo(level=all, keys=['rgb'], to=x, delete_after=True)
NAGAddSelfLoops()
OnTheFlyInstanceGraph(level=-1, num_classes=3, adjacency_mode=radius, k_max=30, radius=0.1, use_batch=True, centroid_mode=iou, centroid_level=1, smooth_affinity=True)

NAG(num_levels=4, num_points=[162752, 37349, 37347, 34403], device=cpu)
NAG(num_levels=4, num_points=[162752, 37349, 37347, 34403], device=cuda:0)
Data(y=[162752, 4], pos=[162752, 3], curvature=[162752, 1], length=[162752, 1], planarity=[162752, 1], pos_offset=[3], rgb=[162752, 3], scattering=[162752, 1], super_index=[162752], volume=[162752, 1], sub=Cluster(num_clusters=162752, num_points=174428, device=cuda:0), obj=InstanceData(num_clusters=162752, num_overlaps=162783, num_obj=12, device=cuda:0), x=[162752, 8], semantic_pred=[162752])

**Comment by drprojects** [comment]
Please provide a minimum reproducible example if you want us to have a look. 

_NB: we assume you have not modified the codebase beyond some config files. We do not have the capacity to provide support for custom implementations_

**Comment by meehirmhatrepy** [comment]
please let me know if this code below will help solving the issue asI haven't changed much in the existing code for kitti360.
dataloader:
    batch_size: 1
trainval: False
pc_tiling: null
xy_tiling: null

# CPU-based train transforms
train_transform: null

# CPU-based val transforms
val_transform: ${datamodule.train_transform}

# CPU-based test transforms
test_transform: ${datamodule.val_transform}

# GPU-based val transforms
on_device_val_transform:

    # Add a `node_size` attribute to all segments, this is needed for
    # segment-wise position normalization with UnitSphereNorm
    - transform: NodeSize

    # Apply sampling transforms first to reduce the number of nodes and
    # edges. These operations are compute-intensive and are the reason
    # why these transforms are not performed on CPU
    # - transform: SampleSubNodes
    #   params:
    #     low: 0
    #     high: 1
    #     n_min: 128
    #     n_max: 256

    # Cast all attributes to either float or long. Doing this only now
    # allows speeding up disk I/O and CPU->GPU transfer
    - transform: NAGCast

    # Compute some horizontal and vertical edges on-the-fly. Those are
    # only computed now since they can be deduced from point and node
    # attributes. Besides, the OnTheFlyHorizontalEdgeFeatures transform
    # takes a trimmed graph as input and doubles its size, creating j->i
    # for each input i->j edge
    - transform: OnTheFlyHorizontalEdgeFeatures
      params:
        keys: ${datamodule.edge_hf}
        use_mean_normal: ${eval:'"normal" in ${datamodule.segment_mean_hf}'}
    - transform: OnTheFlyVerticalEdgeFeatures
      params:
        keys: ${datamodule.v_edge_hf}
        use_mean_normal: ${eval:'"normal" in ${datamodule.segment_mean_hf}'}

    # Move all point and segment features to 'x', except for "rgb", on
    # which we want to apply specific transforms
    - transform: NAGAddKeysTo
      params:
        level: 0
        keys: ${eval:'ListConfig([k for k in ${datamodule.point_hf} if k != "rgb"])'}
        to: 'x'
    - transform: NAGAddKeysTo
      params:
        level: '1+'
        keys: ${eval:'ListConfig([k for k in ${datamodule.segment_hf} if k != "rgb"])'}
        to: 'x'

    # Finally move RGB to node and segment features, if need be
    - transform: NAGAddKeysTo
      params:
        keys: 'rgb'
        to: 'x'
        strict: False

    # Add self-loops in the horizontal graph
    - transform: NAGAddSelfLoops

        
# GPU-based test transforms
on_device_test_transform: ${datamodule.on_device_val_transform}



notebook code:

for t in dataset.on_device_transform.transforms:
    if isinstance(t, NAGAddKeysTo):
        t.delete_after = False
 
nag = dataset[index]

nag = dataset.on_device_transform(nag.to(device_widget.value))

with torch.no_grad():
    output = model(nag)

nag[0].semantic_pred = output.voxel_semantic_pred(super_index=nag[0].super_index)

**Comment by drprojects** [comment]
Please provide a set of files, some data, and instructions for us to integrate those in the official project to reproduce your error. A convenient MRI should minimize the effort on our end.

**Comment by meehirmhatrepy** [comment]
I had mistakes in data pre-processing part of my code. Thank you for the help. Also, apologies for the inconvenience caused.
